### ISSUE TITLE
Implement "Device" concept

### ISSUE THREAD

**Issue by antsar** [home-assistant/architecture#36]
Implement `device` to represent physical devices connected to Home Assistant.

## Background

IoT devices with varied attributes must currently be represented as several different entities in Home Assistant. An atmospheric monitoring unit may have humidity and temperature `sensor`s, some other `binary_sensor`s, perhaps a `light`, and maybe a `switch`. Currently, these are all separate entities in Home Assistant, united only by the common underlying `component`.

## Proposal

Create a new concept in Home Assistant: **`device`**. A `device` may have one or more entities ( `binary_sensor`, `light`, `switch`) related to it.

## Justification

This would improve alignment between Home Assistant code/configuration and the physical devices being represented. This would benefit new users trying to understand Home Assistant.

## Drawbacks

Depending on how this is represented in configuration files, this change is likely to break existing installs.

## Related discussion

Related discussion has occurred in issue #8. Some excerpts:

[balloob commented on Dec 13, 2017:](https://github.com/home-assistant/architecture/issues/8#issuecomment-366076373)
> 
> Comment by armills <s>(he posted in the wrong issue)</s>
> <hr>
> 
> Potentially less disruptive proposal that covers both issues:
> 
> We create a new domain `device`, for which each entity represents a single IoT device. (i.e. 1 multisensor, 1 thermostat, 1 smart bulb, 1 switch, etc.) The attributes on this entity can store all of the metadata. (firmware, brand, etc.) The existing entities can use an attribute `device_id` to associate themselves with this device.
> 
> - Since the attributes on the device entity shouldn't change often, we won't have the problem of logging them on every state change.
> 
> - The frontend could use the device_id to make decisions about how to order/group entities if we wanted to, and we could even display the device entity as its own card with a summary of the different entities.
> 
> - Since entities are only linked to a device, we don't open the door to arbitrary nesting of entities.
> 
> - It's a non-breaking change, and the implementation can be rolled out as platforms gain support. It's just a single new entity/attribute that doesn't break the existing way we do things.
> 

[rytilahti commented on Dec 14, 2017](https://github.com/home-assistant/architecture/issues/8#issuecomment-366076392)
> 
> ```
> One power strip entity (or device as discussed above), consisting of:
>   * Switch[6]
>     - turn_on()
>     - turn_off()
>     - switch_custom_service()
>     - .. and_other_switch_domain_functionalities ..
>     * Sensors:
>           - temperature sensor
>           - voltage sensor
>           - ..
> ```

[cgarwood commented on Feb 23](https://github.com/home-assistant/architecture/issues/8#issuecomment-368017408)
> This would be a major architecture change, but what about creating a single entity per device that contains all of the sensor values/states/attributes, instead of a bunch of separate sensor entities.
> 
> Take a zwave multisensor for example. Currently you get separate sensors for motion, temperature, humidity, luminance, alarm_level, alarm_type, burglar, etc. What if, instead, you just had a single entity that contained all of this detail like such:
> 
> ```
> {
>   "temperature" : {
>     "value" : 71.6
>     "unit_of_measurement" : "F"
>   },
>   "humidity" : {
>     "value" : 52
>     "unit_of_measurement" : "%"
>   }
>   "luminance" : {
>     "value" : 123
>     "unit_of_measurement" : "lux"
>   }
>   "motion" : false
>   "alarm_type" : 22
>   "alarm_level" : 0
>   "burglar" : 8
>   "sourcenodeid" : 0
>   "battery_level" : {
>     "value" : 98
>     "unit_of_measurement" : "%"
>   }
> }
> ```
> 
> You could then access any of those values using dot-notation in automations or elsewhere in the config. For example, living_room_multisensor.motion or living_room_multisensor.battery_level.value
> 
> Everything would be treated the same as a state, no more template sensors to pull out attributes. You would add the different bits to the frontend using the same dot notation, so instead of the current setup of sensor.living_room_multisensor_temperature you'd just include living_room_multisensor.temperature in the group/view.
> 
> I'm not sure how the current domain structure of sensor,binary_sensor,etc would fit in with this architecture.

**Comment by balloob** [comment]
Reviewing the proposal and the quotes here, I must say that the idea of devices has grown on me a lot. 

As a proposal to implement it, I would think of something like this:

- Add a new property `device` to the `Entity` class. Returns an instance of `Device` class. Defaults to `None`
- Device class will have properties manufacturer, model, firmware.
- When an entity is added to an entity platform, check if it has a device, if so, register the device with a Device Registry.

_(this proposal would supersede #26)_

**Comment by dgomes** [comment]
Would we be able to inherit from Devices ? 

See very good candidates:
- Cars
- Personal Weather Stations
- System Monitoring (CPU, RAM, Disk)

**Comment by dgomes** [comment]
Also relevant: https://iot.mozilla.org/wot/#thing-resource

**Comment by balloob** [comment]
To continue on my proposal,  I realized that there are also devices that on their own do not have entities, but still should be there. For example, the Hue integration communicates with a Hub, but the hub itself is not an entity. Components should be able to also register entities.

It's up to components and devices to specify a unique ID. Unique ID will be scoped by component or platform name. So in the case of light.hue, device will be identified within hue (both component and platform).

It's important that when devices get registered, we know if they are from a config entry. We can then start generating this tree:

- config entry for Z-Wave (will enabling component)
  - device: 5-in-1 sensor
     - entity: temp sensor
     - entity: humidity sensor
  - device: Z-Wave stick

**Comment by balloob** [comment]
### Device properties

Required:

 - Component: string (hue, nest, etc)
 - Unique ID: string
 - Manufacturer: string
 - Model: string
 - Connection: string (Wifi, ethernet, Z-Wave, Zigbee, Internet)
 - Config Entry ID: string

Optional:

 - Address: string = (For wifi/ethernet is MAC, so we can hook it up with device tracker etc)
 - Firmware version: string
 - Serial NO: string

**Comment by balloob** [comment]
I wonder how we should deal with things like a TV that can be connected via Plex and via Cast and maybe via a Samsung platform?

Would that mean that the device in that case can be shared among config entries 🤔

**Comment by awarecan** [comment]
Like a TV can implement multiple interfaces, Plex, Cast and Samsung. In that case, your required properties should change to:

- Unique ID: string
- Manufacturer: string
- Model: string
- Connection: list of Connection Profile
  * Profile Type: string (Wifi, ethernet, Z-Wave, Zigbee, Internet)
  * Profile Data: such as MAC, IP, etc
- Interfaces: list of
  * Component: string (hue, nest, etc)
  * Config Entry ID: string

**Comment by balloob** [comment]
Good point. 

Just realized, we should be able to infer the component from the config entry.

**Comment by balloob** [comment]
So made a diagram how the different parts relate to one another.

https://docs.google.com/drawings/d/1-wR8t3L1UMRDv9LskH-rDMus4bnvgxa7OeFI_5UZAZ0/edit?usp=sharing

![devices config entries entities 2](https://user-images.githubusercontent.com/1444314/41627260-93339efe-73ed-11e8-93ba-c6d2f785e2da.png)

**Comment by dgomes** [comment]
"Device can be shared between two components"

In the diagram, this use case is depicted by Device C. Device C connects to two entities belonging to Component A (Example: TV that runs two media_players). 

Now I'm wondering about Device D which connects directly to Component B (example: Hue registers the Hub).

Do we have a use case for a Device that connects directly to TWO Components ?

**Comment by balloob** [comment]
The only thing I can think of is something like a server that is running a bunch of stuff.

**Comment by OttoWinter** [comment]
I've not read through all of this, but @dgomes: For me a use case would be to have an ESP (for esphomelib 😉) exposed as a "device". Because such a microcontroller can have multiple components connected to it. For example an RGB light, a binary sensor and sensor.

The advantage would then be that each "node"/device can then be exposed as a single device and be managed like that.

Not sure if that use case applies here though, as the device concept is as far as I see quite connected to config entries.

**Comment by dgomes** [comment]
@OttoWinter but would it be connected to the component or to the entity ?

I'm fine with any answer, I'm just concern with handling the consistency of all these relationships, especially this that are any-to-any.

**Comment by balloob** [comment]
It's not connected perse to config entries, entities could come from configuration yaml or discovery. Although for discovery we'll have to talk about how to define a device.

I prefer for it to be evolving around config entries because each entry represents authorization to access a service/set of devices and that makes a nice UI.

Wondering now if we should allow IP addresses as temporary identifiers, as Mac is not always accessible

**Comment by awarecan** [comment]
I met a challenge may fit for Device to resolve.

In Nest, it break up to tons of sensors, for each Nest Device, we have a `binary_sensor.online`. So, in case network connection has issue, I want to use this sensor to cutoff other sensors/climate/camera's state update process to avoid flush event loop. 

If we have Device, and it has a property say `online`, then everything will become very easy and clear, we should first check device's online status (no matter it maps to a separate sensor or not) before try to update any other Entitiies's state.

**Comment by balloob** [comment]
The data flows from the integration to the device and entities. An integration provides the info and so doesn't need to check device

**Comment by Adminiuga** [comment]
How do we go about a "technical"  or  "debug" information for a device? For example, in Zigbee world it would be tremendously helpful to have information like: Link Quality Indicator, Last Seen, Neighbors etc, which may change quite frequently, but not necessarily should trigger a state change event. 

Or as another example a device representing a "network interface". it is helpful to have debugging information, like number of packets in/out, number of CRC errors, dropped packets etc.

In other words information that you don't need when everything is working fine, but is really needed during troubleshooting/error reporting

**Comment by dgomes** [comment]
@Adminiuga those sensor entities (as they currently are), which you can associate to a Device.

**Comment by balloob** [comment]
I think that such data is out of scope. Zigbee should just get their own panel like Z-Wave and make these things available. There is no need to add that to the core of HASS.

**Comment by Kane610** [comment]
So I've started playing around with a device registry and there is one main thing Im trying to grasp in this design; how the actual coupling of device entity would work. Is it enough with the device to keep a list of entity ids?

**Comment by awarecan** [comment]
Why no, I would say object reference. entity_id can be changed by user.

**Comment by balloob** [comment]
I would expect it to be linked as follows:

- Each entity has a new property `device`. This property is read when an entity is passed to the Entity Platform's `add_entities` inside `async_setup_platform`/`async_setup_entry`
- The device is processed in the registry, either retrieving an existing registration or creating a new one
- The entity registry will store a device id.

This means that entities that do not have a unique ID cannot be linked to a device but that makes sense. If we cannot figure out if we've seen an entity before, we can't store that data either.

**Comment by Kane610** [comment]
I'd like some feedback on my initial alpha draft of device registry whether I'm in the right ball park and general comments. Also if I've misunderstood any points in the discussion above.

Thanks!

**Comment by Kane610** [comment]
Initial documentation https://github.com/home-assistant/developers.home-assistant/pull/81

It does need comments

**Comment by balloob** [comment]
Yes we should! Closing this as it's considered done.
